### ISSUE TITLE
"create" + disk full makes "check --repair" crash

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbackup/borg#364]
This was found by archiver.test_disk_full.

repos + corresponding tracebacks: https://paste.thinkmo.de/bqyrsTv6#borg-corrupt-repos.tgz

E.g. repo1:
```
$ borg check --repair /tmp/mnt/repo
'check --repair' is an experimental feature that might result in data loss.
Type 'YES' if you understand this and want to continue: YES
Starting repository check...
Segment entry data short read [segment 9, offset 575059]: expected 80786, got 65657 bytes
attempting to recover /tmp/mnt/repo/data/0/9
Local Exception.
Traceback (most recent call last):
  File "/home/tw/w/borg/borg/archiver.py", line 1069, in main
    exit_code = archiver.run(args)
  File "/home/tw/w/borg/borg/archiver.py", line 1023, in run
    return args.func(args)
  File "/home/tw/w/borg/borg/archiver.py", line 99, in do_check
    if repository.check(repair=args.repair):
  File "/home/tw/w/borg/borg/repository.py", line 348, in check
    self.compact_segments()
  File "/home/tw/w/borg/borg/repository.py", line 230, in compact_segments
    assert segments[segment] == 0
AssertionError
```

**Comment by ThomasWaldmann** [comment]
Crash looks same as this one: https://github.com/jborg/attic/issues/166#issuecomment-68614341

**Comment by ThomasWaldmann** [comment]
Also: `borg list` crashes with IntegrityError here:
```
$ borg list /tmp/repo1
Data integrity error
Traceback (most recent call last):
  File "/home/tw/w/borg/borg/archiver.py", line 1069, in main
    exit_code = archiver.run(args)
  File "/home/tw/w/borg/borg/archiver.py", line 1023, in run
    return args.func(args)
  File "/home/tw/w/borg/borg/archiver.py", line 381, in do_list
    manifest, key = Manifest.load(repository)
  File "/home/tw/w/borg/borg/helpers.py", line 108, in load
    cdata = repository.get(cls.MANIFEST_ID)
  File "/home/tw/w/borg/borg/repository.py", line 376, in get
    self.index = self.open_index(self.get_transaction_id())
  File "/home/tw/w/borg/borg/repository.py", line 130, in get_transaction_id
    self.replay_segments(replay_from, segments_transaction_id)
  File "/home/tw/w/borg/borg/repository.py", line 246, in replay_segments
    for tag, key, offset in self.io.iter_objects(segment):
  File "/home/tw/w/borg/borg/repository.py", line 541, in iter_objects
    (TAG_PUT, TAG_DELETE, TAG_COMMIT))
  File "/home/tw/w/borg/borg/repository.py", line 602, in _read
    segment, offset, length, len(data)))
borg.helpers.IntegrityError: Segment entry data short read [segment 9, offset 575059]: expected 80786, got 65657 bytes
```
The reason for this is that the chunk is incomplete, the last chunk that was written into segment 9 could not get written completely because the disk was full. Note: there is a 0 byte segment 10 file.

**Comment by ThomasWaldmann** [comment]
Note: a 2nd `borg check --repair` run on same repo does not crash any more.

**Comment by ThomasWaldmann** [comment]
I removed the corruption label. Found out that there is just a bug in "borg check" initialization (and fixed it).

**Comment by ThomasWaldmann** [comment]
Fixed by PR #366. Can someone please review that?
